lib/charms/prometheus_pushgateway_k8s/v0/pushgateway.py

```diff
@@ -0,0 +1,148 @@
+
+
+## Waiting for Pushgateway
+
+When your charm is deployed but the relation is still not added to the Prometheus Pushgateway
+the metrics could not be sent.
+
+For robustness the charm should only send metrics only after the `pushgateway_available` event
+is received.
+
+The following is a bare charm that holds a flag for when the Pushgateway is available, which is
+set to True when the mentioned event arrives:
+
+```
+class ExampleCharm(CharmBase):
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._stored.set_default(pushgateway_available=False)
+
+        self.ppi = PrometheusPushgatewayInterface(self)
+        self.framework.observe(self.ppi.on.pushgateway_available, self._on_pushgateway_available)
+
+    def _on_pushgateway_available(self, _):
+        self._stored.pushgateway_available = True
+```
+"""
+
+import json
+import logging
+from typing import Union
+
+import requests
+from ops.charm import CharmBase, HookEvent
+from ops.framework import EventBase, EventSource, Object, ObjectEvents, StoredState
+
+logger = logging.getLogger(__name__)
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "a0065690fe484ef296d5847fcbf1d728"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["requests"]
+
+
+class _PushgatewayAvailableEvent(EventBase):
+    """Event emitted when the Prometheus Pushgateway is available to receive metrics."""
+
+
+class _Events(ObjectEvents):
+    """Events for the interface."""
+
+    pushgateway_available = EventSource(_PushgatewayAvailableEvent)
+
+
+class PrometheusPushgatewayInterface(Object):
+    """Interface for the Prometheus Pushgateway."""
+
+    _stored = StoredState()
+    on = _Events()
+
+    def __init__(self, charm: CharmBase):
+        """Construct the interface for the Prometheus Pushgateway.
+
+        Args:
+            charm: a `CharmBase` object that manages this
+                `MetricsEndpointProvider` object. Typically, this is
+                `self` in the instantiating class.
+        """
+        super().__init__(charm, None)
+        self._stored.set_default(pushgateway_url=None)
+
+        self.framework.observe(charm.on.pushgateway_relation_created, self._on_push_relation)
+        self.framework.observe(charm.on.pushgateway_relation_changed, self._on_push_relation)
+
+    def _on_push_relation(self, event: HookEvent) -> None:
+        """Receive the push endpoint information."""
+        raw = event.relation.data[event.app].get("push-endpoint")
+        if raw is not None:
+            logger.info("Received push endpoint information: %r", raw)
+            info = json.loads(raw)
+            self._stored.pushgateway_url = "http://{hostname}:{port}/".format_map(info)
```

Comment:
I didn't understand the first question, I'm not copying anything. I'm building an URL from the data that came in the relation.

We need to "store" the URL because we need it later when the charm is run again.

Why the extra complexity of a property? What are the advantages?

Reply:
If the relation is there and there is data in the relation, we can always fetch that piece of data on-demand and always have the most up-to-date value. 
Unless you need to detect changes to the pushgateway-url, you could do:
```
@property
def _pushgateway_url(self) -> Optional[str]:
    try:
         raw_data = self.model.get_relation('foo').data[event.app].get("push-endpoint")
    except ModelError:
         return None
    try:
         data = json.loads(raw_data)
    except json.JSONDecodeError:
         return None
    return f"http://{data['hostname']}:{data['port']}/"
```

and get rid of stored state. (it's best to avoid using it whenever you can)

Reply:
Now the information is retrieved dynamically as indicated here.